packages/blocks/src/root-block/widgets/element-toolbar/align-button.ts

```diff
@@ -232,24 +323,93 @@ export class EdgelessAlignButton extends WithDisposable(LitElement) {
     });
   }
 
+  private _splitElementsToChunks(models: GfxModel[]) {
+    const sortByCenterX = (a: GfxModel, b: GfxModel) =>
+      a.elementBound.center[0] - b.elementBound.center[0];
+    const sortByCenterY = (a: GfxModel, b: GfxModel) =>
+      a.elementBound.center[1] - b.elementBound.center[1];
+    const elements = models.filter(ele => {
+      if (
+        ele instanceof ConnectorElementModel &&
+        (ele.target.id || ele.source.id)
+      ) {
+        return false;
+      }
+      return true;
+    });
+    elements.sort(sortByCenterY);
+    const chunks = chunk(elements, 4);
+    chunks.forEach(items => items.sort(sortByCenterX));
+    return chunks;
+  }
+
+  private _updatChildElementsXYWH(
+    parent: GroupElementModel | FrameBlockModel,
+    parentBound: Bound
+  ) {
+    const groupBound = Bound.deserialize(parent.xywh);
+    const scaleX = parentBound.w / groupBound.w;
+    const scaleY = parentBound.h / groupBound.h;
+    parent.childElements.forEach(child => {
+      const childBound = Bound.deserialize(child.xywh);
+      childBound.x = parentBound.x + scaleX * (childBound.x - groupBound.x);
+      childBound.y = parentBound.y + scaleY * (childBound.y - groupBound.y);
+      childBound.w = scaleX * childBound.w;
+      childBound.h = scaleY * childBound.h;
+      this._updateXYWH(child, childBound);
+    });
+  }
+
   private _updateXYWH(ele: BlockSuite.EdgelessModel, bound: Bound) {
     if (ele instanceof ConnectorElementModel) {
       ele.moveTo(bound);
     } else if (ele instanceof GroupElementModel) {
-      const groupBound = Bound.deserialize(ele.xywh);
-      ele.childElements.forEach(child => {
-        const newBound = Bound.deserialize(child.xywh);
-        newBound.x += bound.x - groupBound.x;
-        newBound.y += bound.y - groupBound.y;
-        this._updateXYWH(child, newBound);
+      this._updatChildElementsXYWH(ele, bound);
+      this.edgeless.service.updateElement(ele.id, {
+        xywh: bound.serialize(),
+      });
+    } else if (ele instanceof FrameBlockModel) {
```

Comment:
Maybe merge with line 366?

Reply:
You can use `isContainerEle` or `surface.isContainer` to assert a element is container, the you can use  `element.childElements` to merge the three branch

Reply:
@L-Sun Cool. Code updated.